### PR TITLE
Read and write ROS controllers yaml file

### PR DESCRIPTION
**Pull request by MohmadAyman**
### Description

Adding functions to read and write `ros_contollers.yaml` file.
Adding an automated way to add FollowJointTrajectory action controller for each planning group, e.g. 
```
  controller_list:
  - name: g1_controller
    action_ns: follow_joint_trajectory
    default: True
    type: FollowJointTrajectory
    joints:
      - joint1
      - joint2
```

@davetcoleman @mcevoyandy 
### Checklist
- [x] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [needed] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Include a screenshot if changing a GUI
- [ ] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  save_and_add->setFont(save_and_add_font);
+  new_buttons_layout->addWidget(save_and_add);
+
+  QLabel* add_subtitle = new QLabel("Recommended: ", this);
+  QFont add_subtitle_font(QFont().defaultFamily(), 10, QFont::Bold);
+  add_subtitle->setFont(add_subtitle_font);
+  new_buttons_layout->addWidget(add_subtitle);
+
+  // Save and add groups
+  QPushButton* btn_save_groups_joints = new QPushButton("Add Group Joints", this);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: "Add Planning Group"

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+  btn_save_groups_joints->setMaximumWidth(200);
+  connect(btn_save_groups_joints, SIGNAL(clicked()), this, SIGNAL(saveJointsGroups()));
+  new_buttons_layout->addWidget(btn_save_groups_joints);
+
+  QLabel* add_subtitle2 = new QLabel("Advanced Options:", this);
+  add_subtitle2->setFont(add_subtitle_font);
+  new_buttons_layout->addWidget(add_subtitle2);
+
+  // Save and add joints
+  QPushButton* btn_save_joints = new QPushButton("Add Joints", this);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: "Add Individual Joints"

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+// ******************************************************************************************
+// Set the fields with the previous values
+// ******************************************************************************************
+void ControllerEditWidget::setSelected(const std::string& controller_name)
+{
+  controller_name_field_->setText(QString(controller_name.c_str()));
+  Controller* searched_controller_ = config_data_->findControllerByName(controller_name);
+  if (searched_controller_ != NULL)
+  {
+    const std::string controller_type_ = searched_controller_->type_;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: do not use underscores at end

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  this->setLayout(layout);
+}
+
+// ******************************************************************************************
+// Set the fields with the previous values
+// ******************************************************************************************
+void ControllerEditWidget::setSelected(const std::string& controller_name)
+{
+  controller_name_field_->setText(QString(controller_name.c_str()));
+  Controller* searched_controller_ = config_data_->findControllerByName(controller_name);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: do not use underscores at end

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+  if (searched_controller_ != NULL)
+  {
+    const std::string controller_type_ = searched_controller_->type_;
+
+    // Set the controller type combo box
+    int index = controller_type_field_->findText(controller_type_.c_str());
+    if (index == -1)
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: combine line 171 into the if statement

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+  }
+}
+
+// ******************************************************************************************
+// Populate the combo dropdown box with controllers types
+// ******************************************************************************************
+void ControllerEditWidget::loadControllersTypesComboBox()
+{
+  // Only load this combo box once
+  static bool hasLoaded = false;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: has_loaded

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+  if (hasLoaded)
+    return;
+  hasLoaded = true;
+
+  const int types_count = 10;
+  const std::string types[types_count] = { "effort_controllers/JointTrajectoryController",
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: use an std::array here

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+// ******************************************************************************************
+void ControllerEditWidget::loadControllersTypesComboBox()
+{
+  // Only load this combo box once
+  static bool hasLoaded = false;
+  if (hasLoaded)
+    return;
+  hasLoaded = true;
+
+  const int types_count = 10;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: remove this manual counting step

File: moveit_setup_assistant/src/widgets/controller_edit_widget.cpp [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+                                           "velocity_controllers/JointvelocityController" };
+
+  // Remove all old items
+  controller_type_field_->clear();
+
+  // Add FollowJointTrajectory option, the default
+  controller_type_field_->addItem("FollowJointTrajectory");
+
+  // Loop through all controller types and add to combo box
+  for (int i = 0; i < types_count; ++i)
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.cpp]: int --> size_t

File: moveit_setup_assistant/src/widgets/controller_edit_widget.h [hunk 10]
```diff
@@ -0,0 +1,122 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2012, Willow Garage, Inc.
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.h]: change

File: moveit_setup_assistant/src/widgets/controller_edit_widget.h [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  // ******************************************************************************************
+  // Variables
+  // ******************************************************************************************
+
+  /// Contains all the configuration data for the setup assistant
+  moveit_setup_assistant::MoveItConfigDataPtr config_data_;
+
+  // ******************************************************************************************
+  // Private Functions
+  // ******************************************************************************************
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controller_edit_widget.h]: remove this comment

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 12]
```diff
@@ -0,0 +1,834 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2012, Willow Garage, Inc.
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: change

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 13]
```diff
@@ -0,0 +1,834 @@
...
+{
+  // Basic widget container
+  // QWidget* edit_widget = new QWidget(this);
+  QVBoxLayout* layout = new QVBoxLayout();
+
+  layout->setAlignment(Qt::AlignTop);
+
+  // Top Header Area ------------------------------------------------
+  HeaderWidget* header =
+      new HeaderWidget("Controllers settings", "Configuring MoveIt! with the controllers on your robot"
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: Add ". " at end (including space)

Comment (davetcoleman) [reply moveit_setup_assistant/src/widgets/controllers_widget.cpp]: "Configure MoveIt! to use ros_control and integrate with Gazebo"

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 14]
```diff
@@ -0,0 +1,834 @@
...
+
+  layout->addWidget(stacked_layout_widget);
+
+  this->setLayout(layout);
+
+  // Create joint publisher -----------------------------------------
+  ros::NodeHandle nh;
+
+  // Create scene publisher for later use
+  pub_robot_state_ = nh.advertise<moveit_msgs::DisplayRobotState>(MOVEIT_ROBOT_STATE, 1);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: why a publisher? this should be removed....

File: moveit_setup_assistant/src/widgets/controllers_widget.h [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+  std::string current_edit_controller_;
+
+  /// Remember whethere we're editing a controller or adding a new one
+  bool adding_new_controller_;
+
+  /// Contains all the configuration data for the setup assistant
+  moveit_setup_assistant::MoveItConfigDataPtr config_data_;
+
+  /// Remember the publisher for quick publishing later
+  ros::Publisher pub_robot_state_;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.h]: why a publisher? this should be removed....

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 16]
```diff
@@ -0,0 +1,834 @@
...
+  // Main widget
+  QWidget* content_widget = new QWidget(this);
+
+  // Basic widget container
+  QVBoxLayout* layout = new QVBoxLayout(this);
+
+  QHBoxLayout* upper_controls_layout = new QHBoxLayout();
+
+  // Add default controller
+  QPushButton* btn_add_default_ = new QPushButton("&Add Joints Follow Trajectory Controllers", this);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: _ remove

Comment (davetcoleman) [reply moveit_setup_assistant/src/widgets/controllers_widget.cpp]: Rename button: "Auto Add FollowJointTrajectory Controllers For Each Planning Group"

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 17]
```diff
@@ -0,0 +1,834 @@
...
+  btn_add_default_->setMaximumWidth(400);
+  connect(btn_add_default_, SIGNAL(clicked()), this, SLOT(addDefaultControllers()));
+  upper_controls_layout->addWidget(btn_add_default_);
+  upper_controls_layout->setAlignment(btn_add_default_, Qt::AlignRight);
+
+  // Add Controls to layout
+  layout->addLayout(upper_controls_layout);
+
+  // Tree Box ----------------------------------------------------------------------
+  controllers_tree_ = new QTreeWidget(this);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: _

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 18]
```diff
@@ -0,0 +1,834 @@
...
+  // Set the selected joints (right box)
+  joints_widget_->setSelected(this_controller->joints_);
+
+  // Set the title
+  joints_widget_->title_->setText(
+      QString("Edit '").append(QString::fromUtf8(this_controller->name_.c_str())).append("' Joint Collection"));
+
+  // Remember what is currently being edited so we can later save changes
+  current_edit_controller_ = this_controller->name_;
+  // current_edit_element_ = JOINT;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: remove

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 19]
```diff
@@ -0,0 +1,834 @@
...
+
+// ******************************************************************************************
+// Called from Double List widget to highlight joints
+// ******************************************************************************************
+void ControllersWidget::previewSelectedJoints(std::vector<std::string> joints)
+{
+  // Unhighlight all links
+  Q_EMIT unhighlightAll();
+
+  for (int i = 0; i < joints.size(); ++i)
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: size_t

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 20]
```diff
@@ -0,0 +1,834 @@
...
+
+// ******************************************************************************************
+// Called from Double List widget to highlight a group
+// ******************************************************************************************
+void ControllersWidget::previewSelectedGroup(std::vector<std::string> groups)
+{
+  // Unhighlight all links
+  Q_EMIT unhighlightAll();
+
+  for (int i = 0; i < groups.size(); ++i)
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: size_t

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 21]
```diff
@@ -0,0 +1,834 @@
...
+  // Get the user custom properties of the currently selected row
+  if (type_ == 2)
+  {
+    // To which controller does this joint belong
+    controller_item = item->parent()->parent();
+    current_edit_controller_ = controller_item->text(0).toUtf8().constData();
+
+    Controller* this_controller = config_data_->findControllerByName(current_edit_controller_);
+    // Load the data
+    loadJointsScreen(this_controller);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: this is the same as ``type_ ==  1``. Does type 2 apply for this controllers screen? Should it be removed?

File: moveit_setup_assistant/src/widgets/controllers_widget.h [hunk 22]
```diff
@@ -0,0 +1,169 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2012, Willow Garage, Inc.
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.h]: change

File: moveit_setup_assistant/src/widgets/controllers_widget.h [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+  void saveJointsScreen();
+  bool saveControllerScreen();
+  void saveControllerScreenEdit();
+  void saveControllerScreenGroups();
+  void saveJointsGroupsScreen();
+  void cancelEditing();
+  void editSelected();
+
+  /// Called from Double List widget to highlight a joint
+  // void previewClickedJoint( std::string name );
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.h]: remove

File: moveit_setup_assistant/src/widgets/planning_groups_widget.cpp [hunk 24]
```diff
@@ -282,6 +282,7 @@ void PlanningGroupsWidget::loadGroupsTree()
     btn_delete_->hide();
   }
 
+  current_edit_group_.clear();
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/planning_groups_widget.cpp]: these kinds of changes should be in a separate PR... this one is way too big already

File: moveit_setup_assistant/src/widgets/start_screen_widget.cpp [hunk 25]
```diff
@@ -431,6 +431,10 @@ bool StartScreenWidget::loadExistingFiles()
                              .append(kinematics_yaml_path.make_preferred().native().c_str()));
   }
 
+  fs::path controllers_yaml_path = config_data_->config_pkg_path_;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/start_screen_widget.cpp]: add comment explaining what this is

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 26]
```diff
@@ -0,0 +1,834 @@
...
+  // New Controller
+  adding_new_controller_ = true;
+
+  // Load the data
+  loadControllerScreen(NULL);  // NULL indicates this is a new controller, not an existing one
+  changeScreen(2);
+}
+
+// ******************************************************************************************
+// Add a Follow Joint Trajectory action Controller for each Planning Group
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: "Add a FollowJointTrajectory action controller from ros_control for each Planning Group"

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp [hunk 27]
```diff
@@ -0,0 +1,834 @@
...
+// ******************************************************************************************
+void ControllersWidget::addDefaultControllers()
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/controllers_widget.cpp]: can you also, by default, add a joint_state_controller/JointStateController? those are equally important

File: moveit_setup_assistant/src/widgets/planning_groups_widget.cpp (hunk 24 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/widgets/planning_groups_widget.cpp]: I wasn't so sure too, but from the feedback I got from PR where I removed the unused variable, "it is advised not to open too small PR"

Comment (MohmadAyman) [reply moveit_setup_assistant/src/widgets/planning_groups_widget.cpp]: [PR 924](https://github.com/ros-planning/moveit/pull/924)

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 28]
```diff
@@ -590,141 +590,194 @@ bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
...
+              {
+                emitter << *joint_it;
+              }
+              emitter << YAML::EndSeq;
+            }
+            else
+            {
+              emitter << YAML::Value << YAML::BeginMap;
+              emitter << controller_it->joints_[0];
+              emitter << YAML::EndMap;
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: No I can't, because I found that if the `joints` were more than just `1` joint, they have to be written in the YAML as a `seq` and a `map` if only `1`. other than that `ros_control` doesn't read it right.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 29]
```diff
@@ -590,141 +590,194 @@ bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
...
+          emitter << YAML::Value << "follow_joint_trajectory";
+          emitter << YAML::Key << "default";
+          emitter << YAML::Value << "True";
+          emitter << YAML::Key << "type";
+          emitter << YAML::Value << controller_it->type_;
+          // Write joints
+          emitter << YAML::Key << "joints";
+          {
+            joints_count_ = controller_it->joints_.size();
+            if (joints_count_ != 1)
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: same previous loop.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 30]
```diff
@@ -590,141 +590,194 @@ bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
...
+              emitter << YAML::Value << YAML::BeginMap;
+              emitter << controller_it->joints_[0];
+              emitter << YAML::EndMap;
+            }
+          }
+          controllers_config_.erase(controller_it);
+        }
+        else
+        {
+          controller_it++;
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: This loops over all controllers in the `ros_control_config_`, and if we are writing 
 `FollowJointTrajectory` `controller` we write its fields, which are very different than the other controllers.
The rest are supported but the `FollowJointTrajectory` needs to be under `controller_list` namespace, that is why it is in a different loop than the other controllers.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 31]
```diff
@@ -590,141 +590,194 @@ bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
...
+      emitter << YAML::Value << "joint_state_controller/JointStateController";
+      emitter << YAML::Key << "publish_rate";
+      emitter << YAML::Value << "50";
       emitter << YAML::EndMap;
+    }
+    // Write default controllers
+    emitter << YAML::Key << "controller_list" << YAML::Value << YAML::BeginMap;
+    {
+      for (std::vector<Controller>::iterator controller_it = controllers_config_.begin();
+           controller_it != controllers_config_.end();)
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: There is an `erase` in the `if (YES)` and `++controller_it` in the else, doesn't that work?

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp (hunk 17 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/widgets/controllers_widget.cpp]: this is a member variable, shouldnt I leave it.

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp (hunk 21 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/widgets/controllers_widget.cpp]: The type two are the individual `joints`, so this means if what was selected before the edit was a joint, the user should be directed to `joints edit screen`.

File: moveit_setup_assistant/src/widgets/controllers_widget.cpp (hunk 27 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/widgets/controllers_widget.cpp]: Yes I already do that in `moveit_data_config` @ https://github.com/ros-planning/moveit/pull/951/files#diff-b722193e66bd85583ac56b9d68186447R666

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 32]
```diff
@@ -221,6 +231,15 @@ class MoveItConfigData
   /// Email of the author of this config
   std::string author_email_;
 
+  /// Name of the group that holds all the joints to be used by sim contollers
+  std::string controllers_initial_group_;
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: I removed the `sim_controllers` part as it is still to be decided which is the best screen to put them, I put it in the description.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 33]
```diff
@@ -1015,6 +1068,57 @@ bool MoveItConfigData::inputKinematicsYAML(const std::string& file_path)
   return true;  // file created successfully
 }
 
+// ******************************************************************************************
+// Input ros_controllers.yaml file
+// ******************************************************************************************
+bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Since its a `gui` code, Well there is two options here: first the fast one
>Add move the `addDefaultControllers` function to `moveit_data_config`, by that I'll dodge setting up Qtests.

Second
>Setup QT tests

I think I should do the second, it might delay this PR a bit, but setting up QT tests will come in handy later with other parts of the `setup_assistant`.

[the remark this answers is not in the document]
Comment (davetcoleman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: The moveit_config_data class does (should) not use QT, its in the /tools directory, not in the /widgets directory. It should be a set of easily testable libraries. We don't need QT tests, that's sounds too complicated

File: moveit_setup_assistant/src/widgets/planning_groups_widget.cpp (hunk 24 above)
Comment (davetcoleman) [reply moveit_setup_assistant/src/widgets/planning_groups_widget.cpp]: Yes there's a middle ground :-)

One line change is a bit small, but many hundreds of lines is getting large.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 33 above)
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Okay I'll look into testing it without having to use the QT test, or the moving anything to the `moveit_config_data`.

File: moveit_setup_assistant/src/widgets/setup_assistant_widget.h [hunk 34]
```diff
@@ -220,6 +221,7 @@ private Q_SLOTS:
   EndEffectorsWidget* efw_;
   VirtualJointsWidget* vjw_;
   PassiveJointsWidget* pjw_;
+  ControllersWidget* cw_;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/setup_assistant_widget.h]: these abbreviated variable names are bad practice, but they are my fault 7 years ago when I was a younger programmer. can you add a TODO to make them not abbreviated? I don't want to make this PR any bigger itself.

**Comment by MohmadAyman** [comment]
As suggested by @davecoleman  this PR now does only deal with reading and writing the `ros_controllers.yaml` file.

File: moveit_setup_assistant/test/moveit_config_data_test.cpp [hunk 35]
```diff
@@ -0,0 +1,137 @@
...
+    // boost::filesystem::remove("ros_controllers.yaml");
+  }
+
+protected:
+  urdf::ModelInterfaceSharedPtr urdf_model;
+  srdf::ModelSharedPtr srdf_model;
+  moveit::core::RobotModelConstPtr robot_model;
+};
+
+TEST(MoveItConfigData, ReadingControllers)
```
Comment (MohmadAyman) [inline moveit_setup_assistant/test/moveit_config_data_test.cpp]: The test is still not completely implemented, I've a problem setting the `robot_model` in `config_data`. I've spend some time trying to figure out how to set it in the test, so for efficiency I thought I would ask for ideas here.

File: moveit_setup_assistant/CMakeLists.txt [hunk 36]
```diff
@@ -141,3 +141,15 @@ install(DIRECTORY include/ DESTINATION include)
 install(DIRECTORY launch DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
 install(DIRECTORY resources DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
 install(DIRECTORY templates DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
+
+# Unit tests
+if(CATKIN_ENABLE_TESTING)
+  find_package(moveit_resources REQUIRED)
+  include_directories(${moveit_resources_INCLUDE_DIRS})
+
+  catkin_add_gtest(test_writting_ros_controllers test/moveit_config_data_test.cpp)
```
Comment (davetcoleman) [inline moveit_setup_assistant/CMakeLists.txt]: "writing" (misspelled)

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 37]
```diff
@@ -459,107 +614,6 @@ std::vector<OMPLPlannerDescription> MoveItConfigData::getOMPLPlanners()
   return planner_des;
 }
 
-// ******************************************************************************************
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: just curious - why move these files? it makes reviewing harder because i can't tell what has actually changed

File: moveit_setup_assistant/test/moveit_config_data_test.cpp [hunk 38]
```diff
@@ -0,0 +1,137 @@
...
+#include <iostream>  // For writing yaml and launch files
+#include <urdf_parser/urdf_parser.h>
+#include <fstream>
+#include <string>
+#include <boost/filesystem/path.hpp>
+#include <moveit_resources/config.h>
+#include <boost/filesystem.hpp>  // for creating folders/files
+#include <moveit/setup_assistant/tools/moveit_config_data.h>
+
+// #include <moveit/setup_assistant/src/widgets/controllers_widget.h>
```
Comment (davetcoleman) [inline moveit_setup_assistant/test/moveit_config_data_test.cpp]: remove

File: moveit_setup_assistant/test/moveit_config_data_test.cpp [hunk 39]
```diff
@@ -0,0 +1,137 @@
...
+
+// This tests adding FollowJointTrajectory controllers for each planning group to ros_controller.yaml
```
Comment (davetcoleman) [inline moveit_setup_assistant/test/moveit_config_data_test.cpp]: "This test adds..."

File: moveit_setup_assistant/test/moveit_config_data_test.cpp [hunk 40]
```diff
@@ -0,0 +1,137 @@
...
+class MoveItConfigData : public testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    boost::filesystem::path res_path(MOVEIT_TEST_RESOURCES_DIR);
+
+    srdf_model.reset(new srdf::Model());
+    std::string xml_string;
+    std::fstream xml_file((res_path / "pr2_description/urdf/robot.xml").string().c_str(), std::fstream::in);
```
Comment (davetcoleman) [inline moveit_setup_assistant/test/moveit_config_data_test.cpp]: lets move away from using the PR2 but rather the Franka Emika robot instead. we might need to add Franka Emika into moveit_resources. @mlautman any opinions?

File: moveit_setup_assistant/test/moveit_config_data_test.cpp [hunk 41]
```diff
@@ -0,0 +1,137 @@
...
+  // Initially no controllers
+  EXPECT_EQ(config_data_->ros_controllers_config_.size(), 0);
+
+  // Setup the robot model
+  // config_data_->urdf_model_ = urdf::ModelPtr;
+
+  // Read ros_controllers.yaml
+  // config_data_->inputROSControllersYAML("ros_controllers.yaml");
+
+  // Verify that we read the file correctly
```
Comment (davetcoleman) [inline moveit_setup_assistant/test/moveit_config_data_test.cpp]: seems to be on the right track!

File: moveit_setup_assistant/test/moveit_config_data_test.cpp (hunk 39 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/test/moveit_config_data_test.cpp]: This tests adding  `FollowJointTrajectory` controllers, isn't that Okay?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 37 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Sorry I didn't understand, which files?

Comment (davetcoleman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: oops i mean functions.... why move the functions to a different location in the file?

File: moveit_setup_assistant/test/moveit_config_data_test.cpp (hunk 40 above)
Comment (mlautman) [reply moveit_setup_assistant/test/moveit_config_data_test.cpp]: I'm strongly in favor of replacing the PR2 with the Panda wherever possible. @MohmadAyman a small side project that would be a huge improvement would be to add the Panda to `moveit_resources` so that it can be used in this test.

Comment (MohmadAyman) [reply moveit_setup_assistant/test/moveit_config_data_test.cpp]: Sure, maybe will add it in a separate PR.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 42]
```diff
@@ -274,6 +282,107 @@ bool MoveItConfigData::outputOMPLPlanningYAML(const std::string& file_path)
   return true;  // file created successfully
 }
 
+// ******************************************************************************************
+// Output kinematic config files
+// ******************************************************************************************
+bool MoveItConfigData::outputKinematicsYAML(const std::string& file_path)
```
Comment (MohmadAyman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: @davetcoleman 
I didnt change anything in these functions, does git sees that as a diff because I've added the `setRobotModel` function above?

**Comment by MohmadAyman** [comment]
Ping @davetcoleman @mcevoyandy 
are there any further comments concerning this PR?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 43]
```diff
@@ -78,6 +77,15 @@ MoveItConfigData::~MoveItConfigData()
 {
 }
 
+// ******************************************************************************************
+// Load a robot model
+// ******************************************************************************************
+void MoveItConfigData::setRobotModel(robot_model::RobotModelPtr robot_model)
+{
+  robot_model_ = robot_model;
+  robot_model_const_ = robot_model;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: just curious: do you know why i originally had to add a const and non-const version of robot_model? this is very hacky. shame on me (~6 years ago)

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 42 above)
Comment (davetcoleman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: not sure what you mean 'diff'?

**Comment by davetcoleman** [comment]
@mcevoyandy please review

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 42 above)
Comment (mcevoyandy) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Looks like you moved them up in the file? I think this is why git sees this as a large diff.

**Comment by mcevoyandy** [comment]
@davetcoleman lgtm.

**Comment by davetcoleman** [comment]
Great job @MohmadAyman !!

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 42 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Sorry just saw the comment, I mean the difference that `git` saw.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 43 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Well the `robot_model_const_` is only used in `getRobotModel`, while `robot_model_` is used whenever we are to edit sth in the robot model.

We can remove `robot_model_const_` and just `const_cast` the `robot_model_` in the `getRobotModel()`, does that sound better?

**Comment by MohmadAyman** [comment]
Thank you :smiley_cat:
